presto-spi/src/main/java/com/facebook/presto/spi/block/ByteArrayBlock.java

```diff
@@ -199,6 +200,28 @@ public Block copyRegion(int positionOffset, int length)
         return new ByteArrayBlock(0, length, newValueIsNull, newValues);
     }
 
+    @Override
+    public BlockLease flatten(ArrayAllocator arrayAllocator)
+    {
+        if (arrayOffset != 0) {
+            byte[] bytes = arrayAllocator.getByteArray(positionCount);
+            System.arraycopy(values, arrayOffset, bytes, 0, positionCount);
+            if (valueIsNull == null) {
+                ByteArrayBlock byteArrayBlock = new ByteArrayBlock(positionCount, Optional.empty(), bytes);
+                return newLease(bytes, byteArrayBlock, arrayAllocator);
+            }
+            else {
+                boolean[] isNull = arrayAllocator.getBooleanArray(positionCount);
+                System.arraycopy(valueIsNull, arrayOffset, isNull, 0, positionCount);
+                ByteArrayBlock byteArrayBlock = new ByteArrayBlock(0, positionCount, isNull, bytes);
+                return newLease(bytes, isNull, byteArrayBlock, arrayAllocator);
+            }
+        }
```

Comment:
Is the purpose of this code block if (arrayOffset != 0) {...} just to hoist out the arrayOffset? Is the memory copy cost compensated by the gain of removing arrayOffset? For repartitioning I didn't observe much benefit removing the arrayOffset. Do we have to do memory copies? Can there be other ways?

Reply:
For joins and expressions I found the arrayOffset cost was significant.  If you're comparing this with the prior work, it's basically the same cost, because before we would loop through the map and increment the map + arrayOffset. Here we shift the values by amount arrayOffset, but the memory copy is an intrinsic and I expect it's cheaper.